android/jni/com/mapswithme/maps/SearchEngine.cpp

```diff
@@ -180,16 +364,19 @@ extern "C"
     g_mapResultsMethod = jni::GetMethodID(env, g_javaListener, "onMapSearchResults", "([Lcom/mapswithme/maps/search/NativeMapSearchListener$Result;JZ)V");
     g_mapResultClass = jni::GetGlobalClassRef(env, "com/mapswithme/maps/search/NativeMapSearchListener$Result");
     g_mapResultCtor = jni::GetConstructorID(env, g_mapResultClass, "(Ljava/lang/String;Ljava/lang/String;)V");
+
+    g_hotelsFilterBuilder.Init(env);
   }
 
   JNIEXPORT jboolean JNICALL
   Java_com_mapswithme_maps_search_SearchEngine_nativeRunSearch(JNIEnv * env, jclass clazz, jbyteArray bytes, jstring lang,
-                                                               jlong timestamp, jboolean hasPosition, jdouble lat, jdouble lon)
+                                                               jlong timestamp, jboolean hasPosition, jdouble lat, jdouble lon, jobject hotelsFilter)
   {
     search::EverywhereSearchParams params;
     params.m_query = jni::ToNativeString(env, bytes);
     params.m_inputLocale = ReplaceDeprecatedLanguageCode(jni::ToNativeString(env, lang));
     params.m_onResults = bind(&OnResults, _1, timestamp, false, hasPosition, lat, lon);
+    params.m_hotelsFilter = g_hotelsFilterBuilder.Build(env, hotelsFilter);
```

Comment:
The `hotelsFilter` is a local reference, and this reference is valid only duration the method call according to JNI specification - http://docs.oracle.com/javase/7/docs/technotes/guides/jni/spec/design.html#wp1242 . That's why, I'm pretty sure this reference should be converted to global one first (`env->NewGlobalRef(env, hotelsFilter)`) and then it can be passed in to the core code. Also, a good practise is to delete this global reference when it's no needed anymore (env->DeleteGlobalRef(g_hotelsFilter)) or if we know that this object should be in memory during application live we can skip deletion I guess.

Probably I missed something and wrong about invalid reference to hotelsFilter, but it looks to me so so far. Let me know if I'm wrong